pkg/client/unversioned/portforward/portforward.go

```diff
@@ -250,67 +254,79 @@ func (pf *PortForwarder) waitForConnection(listener net.Listener, port Forwarded
 func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 	defer conn.Close()
 
-	glog.Infof("Handling connection for %d", port.Local)
+	pf.connectionIDLock.Lock()
+	connectionID := pf.nextConnectionID
+	pf.nextConnectionID++
+	pf.connectionIDLock.Unlock()
 
-	errorChan := make(chan error)
-	doneChan := make(chan struct{}, 2)
+	glog.Infof("Handling connection for %d", port.Local)
 
 	// create error stream
 	headers := http.Header{}
 	headers.Set(api.StreamType, api.StreamTypeError)
 	headers.Set(api.PortHeader, fmt.Sprintf("%d", port.Remote))
+	headers.Set(api.ConnectionIDHeader, strconv.Itoa(connectionID))
 	errorStream, err := pf.streamConn.CreateStream(headers)
 	if err != nil {
-		glog.Errorf("Error creating error stream for port %d -> %d: %v", port.Local, port.Remote, err)
+		util.HandleError(fmt.Errorf("error creating error stream for port %d -> %d: %v", port.Local, port.Remote, err))
 		return
 	}
-	defer errorStream.Reset()
+	// we're not writing to this stream
+	errorStream.Close()
+
+	errorChan := make(chan error)
 	go func() {
 		message, err := ioutil.ReadAll(errorStream)
-		if err != nil && err != io.EOF {
-			errorChan <- fmt.Errorf("Error reading from error stream for port %d -> %d: %v", port.Local, port.Remote, err)
-		}
-		if len(message) > 0 {
-			errorChan <- fmt.Errorf("An error occurred forwarding %d -> %d: %v", port.Local, port.Remote, string(message))
+		switch {
+		case err != nil:
+			errorChan <- fmt.Errorf("error reading from error stream for port %d -> %d: %v", port.Local, port.Remote, err)
+		case len(message) > 0:
+			errorChan <- fmt.Errorf("an error occurred forwarding %d -> %d: %v", port.Local, port.Remote, string(message))
+		default:
+			errorChan <- nil
```

Comment:
Why write a nil value to the channel when you're just going to close it? Reading from a closed channel returns the default value (nil here).